### DISCUSSION TITLE
UIKit `NavigateAndLoad` does not work

### DISCUSSION THREAD

**Discussion by havilog** [pointfreeco/swift-composable-architecture#1806] [Q&A] [answered]
I'm using TCA with UIKit.

my child reducer doesn't receive any action from view.

so i've tried it with official UIKitCaseStudies - NavigateAndLoad (EagerNavigation Reducer) on `main` branch, 

and it still doesn't work.

is it TCA bug?

**Comment by tgrapperon** [comment, marked as answer]
There likely was an omission when converting this study to the `Reducer` protocol. I've submitted a PR to fix this in #1807. It consists in making sure that you embed your child reducer into the parent, here with an `ifLet` reducer modifier. 
I'm not sure this is what's happening in your case, but forgetting to embed a child into a parent (with `ifLet`, `forEach`, `Scope`, …) is usually the principal reason for unresponsive or "dead" child features.

**Reply by havilog** [reply]
@tgrapperon 

Unfortunately, it wasn't my case.
in my case, I made my parent reducer as follows.

```swift
    public var body: some ReducerProtocol<State, Action> {
        Reduce(reduce)
            .ifLet(\.optionalChild, action: /Action.optionalChild) { 
                Child()
            }
    }
    
    public func reduce(into state: inout State, action: Action) -> EffectTask<Action> {
        switch action {
        case .something:
            state.optionalChild = .init()
            return .none
        }
    }
```

my child reducer has not received any action, only the parent reducer received this action.

so I tried like this.

```swift
    public var body: some ReducerProtocol<State, Action> {
        Reduce { state, action in
            switch action {
            case .something:
                state.optionalChild = .init()
                return .none
            }
        }
        .ifLet(\.optionalChild, action: /Action.optionalChild) { 
            Child()
        }
    }
```

and it works.

what is the differance between `Reduce(reduce)` + `ifLet` and `Reduce { ... }` + `ifLet` ?

**Reply by tgrapperon** [reply]
The problem in you case was that you defined both `body` and `reduce`, and this is something that is not supported. You have to pick one of the styles, otherwise only `reduce` will be executed.
The thing you tried last is the way you should do. If you want to keep the "core" reducer's implementation out of `body`, (in the same style you used when it wasn't working), it could also work if you rename the `reduce` function differently. For example if your rename it `core` or something like this:
```swift
public func core(into state: inout State, action: Action) -> EffectTask<Action> { … }
public var body: some ReducerProtocol<State, Action> {
  Reduce(core)
    .ifLet(\.optionalChild, action: /Action.optionalChild) { 
      Child()
    }
  }
```
This approach is often used if you have many domains with many `ifLet`, `forEach` or `Scope`, to make the structure more visible to you. It doesn't change the way it works compared to `body` alone, provided you don't name it `reduce`!

So to sum up, you can use `body`, or you can use `reduce`. In the latter case, if you also implement a `body`, `body` will not be executed (which is what was happening). If you want more insight, the thing that is running body is the default implementation of `reduce` in `ReducerProtocol`, if you define a `reduce` function, you override this default implementation, and `body` is not used.

I wonder if one wouldn't be able to install safeguards about this easy mistake. Xcode doesn't help neither, because it autocompletes very easily the `reduce` function.

**Reply by havilog** [reply]
thanks bro. I totally understand :)
